### ISSUE TITLE
support full width character

### ISSUE THREAD

**Issue by disjukr** [slap-editor/slap#107]
for CJK users
http://en.wikipedia.org/wiki/Halfwidth_and_fullwidth_forms

**Comment by azu** [comment]
:+1: 
well-known as East Asian Width(UTF-8 ambiguous) problem.
[Emojis is also that](https://gist.github.com/mandel59/d05522cdb42a652bf423 "Emoji").

**Comment by secrettriangle** [comment]
So what's the actual problem? Can you provide me with a file I can look at?

**Comment by chjj** [comment]
This is a problem with blessed's screen buffer. It assumes all characters are one cell. An assumption that's true a lot of the time. Double width characters are really annoying - they make everything slower even if you're not using them. I have to find an efficient way to implement them.

**Comment by chjj** [comment]
Also, a lot of terminals aren't even compiled with double-width support, which means even if blessed keeps up its end of the deal, the terminal still might fail and produce even stranger effects. It sounds awful, but I sometimes wish double-width characters would just die. They don't belong on typewriters. I'm sorry, east asian languages. I'll still look for an elegant and efficient solution, but I'll probably complain while doing it.

**Comment by chjj** [comment]
@secrettriangle, @azu, @disjukr, double width characters can be handled by screen rendering and parseContent now. Maybe some revisions need to happen inside slap now though.

**Comment by chjj** [comment]
Double width chars and surrogate pairs moved behind the `fullUnicode` option.

**Comment by chjj** [comment]
Just a summary of this implementation, and a few things I figured out while implementing this:

VTE (gnome-terminal, etc) completely breaks when trying to display a character which is both double-width and a surrogate pair. If it's just displaying a single-width surrogate pair, the character will actually EAT the character after it when rendered. This is just VTE's misbehavior. There is no way for blessed to fix it.

So far the only 2 terminals that can properly render double-width chars, surrogate doubles, and surrogate singles are: rxvt-unicode and xterm (although xterm can't display the unicode glyphs themselves, it does _not_ get the spacing mixed up like other terminals).

This means that there is, so far, only one terminal on earth that can take full advantage of this feature (rxvt-unicode). Now when you take that into account along with the percentage of people who use double-width chars and surrogate pairs in a terminal, you're left with an arguably minimal fraction of the terminal-using population. These factors make this a slightly obscure feature.

I suspected these things originally, which is why I was hesitant to implement this. I knew for a fact most terminals couldn't render double-widths properly anyway, so what is the point of the terminal app sacrificing performance and doing backflips over string length inconsistencies to allow double-width chars?

For these reasons, this feature will most likely stay behind the `fullUnicode` option. It's cool to have, and I think it's as performant as it can be (much faster than I thought I could make it originally). Maybe one day VTE and other terminals will start pulling their weight and implement these things correctly.

In short, if you want to use `fullUnicode` use xterm or, ideally, rxvt-unicode.

All that being said, I think it's a good idea to have this feature enabled by default for a text editor. The user could disable it by editing the config file if they decided to use a subpar terminal.

**Comment by secrettriangle** [comment]
Awesome work @chjj, thanks so much! :metal:

This will be fixed in an upcoming version of slap.

**Comment by chjj** [comment]
I still haven't tested it on Terminal.app or iTerm2. Hopefully they're better than VTE. It might make this more worthwhile.

**Comment by secrettriangle** [comment]
I have decided to leave `fullUnicode` off by default, as slap is a programmer's text editor and 1) needs to be as fast as possible 2) multiwidth characters are very infrequently used.

However, this functionality is now available. To test, add the following to `~/.slap/config`:
```
[slap]
fullUnicode = true
```

I am closing this ticket. Thanks again, @chjj!

**Comment by chjj** [comment]
@secrettriangle, I just realized there will also be more work for you because of this. For example, I had to update the textarea backspacing, from:

``` js
  } else if (key.name === 'backspace') {
    if (this.value.length) {
      this.value = this.value.slice(0, -1);
    }
```

To:

``` js
  } else if (key.name === 'backspace') {
    if (this.value.length) {
      if (this.screen.fullUnicode) {
        if (this.value[this.value.length - 2]
            && unicode.isSurrogate(this.value[this.value.length - 2])) {
          this.value = this.value.slice(0, -2);
        } else {
          this.value = this.value.slice(0, -1);
        }
      } else {
        this.value = this.value.slice(0, -1);
      }
    }
```

Since surrogate pairs take up an actual `.length` of `2` for each character.

**Comment by chjj** [comment]
Oh, and unicode combining characters will be just as tricky since they're 0-width, however, it might be useful for an editor to allow backspacing a combining character to remove the accent/whathaveyou.

**Comment by chjj** [comment]
And I forgot that a lot of unicode combining characters are surrogate pairs. So, they're 2 in length and 0 in width. Are we liking utf-16 yet?
